go/chat/attachment.go

```diff
@@ -0,0 +1,282 @@
+	region := aws.Region{
+		Name:       asset.Region,
+		S3Endpoint: asset.Endpoint,
+	}
+	conn := s3.New(signer, region)
+	conn.AccessKey = params.AccessKey
+
+	b := conn.Bucket(asset.Bucket)
+
+	body, err := b.GetReader(asset.Path)
+	defer func() {
+		if body != nil {
+			body.Close()
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// to keep track of download progress
+	progWriter := newProgressWriter(progress, asset.Size)
+	tee := io.TeeReader(body, progWriter)
+
+	// decrypt body
+	dec := NewPassThrough()
+	decBody, err := dec.Decrypt(tee, asset.Key)
+	if err != nil {
+		return err
+	}
+
+	n, err := io.Copy(w, decBody)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("downloaded %d bytes", n)
+
+	return nil
+}
+
+// putSingle uploads data in r to S3 with the Put API.  It has to be
+// used for anything less than 5MB.  It can be used for anything up
+// to 5GB, but putMultiPipeline best for anything over 5MB.
+func putSingle(ctx context.Context, log logger.Logger, r io.Reader, size int64, params chat1.S3Params, b *s3.Bucket, progress ProgressReporter) error {
+	log.Debug("s3 putSingle (size = %d)", size)
+
+	// In order to be able to retry the upload, need to read in the entire
+	// attachment.  But putSingle is only called for attachments <= 5MB, so
+	// this isn't horrible.
+	buf := make([]byte, size)
+	n, err := r.Read(buf)
+	if err != nil {
+		return err
+	}
+	if int64(n) != size {
+		return fmt.Errorf("invalid read attachment size: %d (expected %d)", n, size)
+	}
+	sr := bytes.NewReader(buf)
+
+	var lastErr error
+	for i := 0; i < 10; i++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(BackoffDefault.Duration(i)):
+		}
+		log.Debug("s3 putSingle attempt %d", i+1)
+		err := b.PutReader(params.ObjectKey, sr, size, "application/octet-stream", s3.ACL(params.Acl), s3.Options{})
+		if err == nil {
+			log.Debug("putSingle attempt %d success", i+1)
+			return nil
+		}
+		log.Debug("putSingle attempt %d error: %s", i+1, err)
+		lastErr = err
+
+		// move back to beginning of sr buffer for retry
+		sr.Seek(0, io.SeekStart)
+
+	}
+	return fmt.Errorf("failed putSingle (last error: %s)", lastErr)
+}
+
+// putMultiPipeline uploads data in r to S3 using the Multi API.  It uses a
+// pipeline to upload 10 blocks of data concurrently.  Each block is 5MB.
+func putMultiPipeline(ctx context.Context, log logger.Logger, r io.Reader, size int64, params chat1.S3Params, b *s3.Bucket, progress ProgressReporter) error {
+	log.Debug("s3 putMultiPipeline (size = %d)", size)
+	multi, err := b.InitMulti(params.ObjectKey, "application/octet-stream", s3.ACL(params.Acl))
+	if err != nil {
+		log.Debug("InitMulti error: %s", err)
+		return err
+	}
+
+	type job struct {
+		block []byte
+		index int
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	blockCh := make(chan job)
+	retCh := make(chan s3.Part)
+	eg.Go(func() error {
```

Comment:
Yeah, and if you watch the logs it really does help as nothing is held up by a single part upload being slow.  You'll see sometimes part 16 finishing before part 4, for example.